nifi-toolkit/nifi-toolkit-admin/src/main/groovy/org/apache/nifi/toolkit/admin/client/NiFiClientFactory.groovy

```diff
@@ -0,0 +1,159 @@
+            final String truststorePassword = niFiProperties.getProperty(NiFiProperties.SECURITY_TRUSTSTORE_PASSWD);
+
+            if(keystore.startsWith("./")){
+                keystore = keystore.replace("./",nifiInstallDir+"/")
+            }
+            if(truststore.startsWith("./")){
+                truststore = truststore.replace("./",nifiInstallDir+"/")
+            }
+
+            sslContext = createSslContext(
+                    keystore.trim(),
+                    keystorePassword.trim().toCharArray(),
+                    keystoreType.trim(),
+                    truststore.trim(),
+                    truststorePassword.trim().toCharArray(),
+                    truststoreType.trim(),
+                    "TLS");
+        }
+
+        final ClientConfig config = new DefaultClientConfig();
+
+        if (sslContext != null) {
+            config.getProperties().put(HTTPSProperties.PROPERTY_HTTPS_PROPERTIES,new HTTPSProperties(new NiFiHostnameVerifier(), sslContext));
+        }
+
+        return  Client.create(config);
+
+    }
+
+
+    private static SSLContext createSslContext(
+            final String keystore, final char[] keystorePasswd, final String keystoreType,
+            final String truststore, final char[] truststorePasswd, final String truststoreType,
+            final String protocol)
+            throws KeyStoreException, IOException, NoSuchAlgorithmException, CertificateException,
+                    UnrecoverableKeyException, KeyManagementException {
+
+        // prepare the keystore
+        final KeyStore keyStore = KeyStore.getInstance(keystoreType);
+        final InputStream keyStoreStream = new FileInputStream(keystore)
+            keyStore.load(keyStoreStream, keystorePasswd);
+
+
+        final KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+        keyManagerFactory.init(keyStore, keystorePasswd);
+
+        // prepare the truststore
+        final KeyStore trustStore = KeyStore.getInstance(truststoreType);
+        final InputStream trustStoreStream = new FileInputStream(truststore)
+        trustStore.load(trustStoreStream, truststorePasswd);
+
+        final TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+        trustManagerFactory.init(trustStore);
+
+        // initialize the ssl context
+        final SSLContext sslContext = SSLContext.getInstance(protocol);
+        sslContext.init(keyManagerFactory.getKeyManagers(), trustManagerFactory.getTrustManagers(), new SecureRandom());
+        return sslContext;
+    }
+
+    private static class NiFiHostnameVerifier implements HostnameVerifier {
+
+        @Override
+        public boolean verify(final String hostname, final SSLSession ssls) {
+            try {
+                for (final Certificate peerCertificate : ssls.getPeerCertificates()) {
```

Comment:
I believe this is too lenient in that it iterates over the entire certificate chain presented by the client. We are only interested in the leaf/final certificate which is the actual peer -- this should be `ssls.getPeerCertificates()[0]` (with proper edge checking, of course).